polynote-kernel/src/main/scala/polynote/kernel/dependency/IvyFetcher.scala

```diff
@@ -0,0 +1,306 @@
+package polynote.kernel.dependency
+
+import java.io.File
+import java.net.URI
+import java.nio.file.Path
+import java.util.concurrent.Executors
+import java.util.concurrent.atomic.AtomicInteger
+
+import cats.effect.{ContextShift, IO}
+import cats.syntax.apply._
+import cats.syntax.parallel._
+import cats.syntax.either._
+import cats.syntax.traverse._
+import cats.instances.list._
+import cats.instances.either._
+import polynote.config.{DependencyConfigs, RepositoryConfig, ivy, maven}
+import polynote.kernel.{KernelStatusUpdate, TaskInfo, TaskStatus, UpdatedTasks}
+import polynote.kernel.util.Publish
+import org.apache.ivy.Ivy
+import org.apache.ivy.core.cache.{CacheDownloadOptions, RepositoryCacheManager}
+import org.apache.ivy.core.module.descriptor.{Artifact, DefaultDependencyDescriptor, DefaultModuleDescriptor}
+import org.apache.ivy.core.module.id.ModuleRevisionId
+import org.apache.ivy.core.report.{ArtifactDownloadReport, DownloadReport, DownloadStatus}
+import org.apache.ivy.core.resolve.{DownloadOptions, ResolveOptions}
+import org.apache.ivy.core.settings.IvySettings
+import org.apache.ivy.plugins.namespace.NameSpaceHelper
+import org.apache.ivy.plugins.repository.file.FileResource
+import org.apache.ivy.plugins.repository.url.URLResource
+import org.apache.ivy.plugins.repository.{ArtifactResourceResolver, Resource, ResourceDownloader, TransferEvent, TransferListener}
+import org.apache.ivy.plugins.resolver.util.ResolvedResource
+import org.apache.ivy.plugins.resolver.{CacheResolver, ChainResolver, IBiblioResolver, URLResolver}
+import org.apache.ivy.util.filter.{Filter => IvyFilter}
+import org.log4s.getLogger
+
+import scala.concurrent.ExecutionContext
+import scala.collection.JavaConverters._
+
+class IvyFetcher extends URLDependencyFetcher {
```

Comment:
This fetcher involves a lot of ugly crap because it is calling all these ugly Java APIs.

We'll probably delete it soon, but I'd like to leave it in place for now in case we want to mess with it some more. In general Ivy seems a bit more robust than Coursier in some ways (though Coursier improved dramatically in the last few milestone releases).